### PR TITLE
Unit tests for KindMapping.h

### PR DESCRIPTION
**Pull request by Sameeranjoshi**
Please suggest changes If I missed some parts.

### CODE REVIEW

File: flang/unittests/Optimizer/KindMappingTest.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+TEST_F(DefaultStringTests, getCharacterBitsizeTest) {
+  EXPECT_EQ(defaultString->getCharacterBitsize(10), (unsigned)80);
+  EXPECT_EQ(defaultString->getCharacterBitsize(0), (unsigned)0);
+}
+
+TEST_F(DefaultStringTests, getLogicalBitsizeTest) {
+  EXPECT_EQ(defaultString->getLogicalBitsize(10), (unsigned)80);
+  // Unsigned values are expected
+  std::string actual = std::to_string(defaultString->getLogicalBitsize(-10));
+  std::string expect("-80");
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/KindMappingTest.cpp:66]: Why is this test different than the others?

File: flang/unittests/Optimizer/KindMappingTest.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+TEST_F(DefaultStringTests, getRealTypeIDTest) {
+  EXPECT_EQ(defaultString->getRealTypeID(2), LLVMTypeID::HalfTyID);
+  EXPECT_EQ(defaultString->getRealTypeID(4), LLVMTypeID::FloatTyID);
+  EXPECT_EQ(defaultString->getRealTypeID(8), LLVMTypeID::DoubleTyID);
+  EXPECT_EQ(defaultString->getRealTypeID(10), LLVMTypeID::X86_FP80TyID);
+  EXPECT_EQ(defaultString->getRealTypeID(16), LLVMTypeID::FP128TyID);
+  // Default cases
+  EXPECT_EQ(defaultString->getRealTypeID(-1), LLVMTypeID::FloatTyID);
+  EXPECT_EQ(defaultString->getRealTypeID(1), LLVMTypeID::FloatTyID);
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/KindMappingTest.cpp:78]: +1

**Comment by Sameeranjoshi** [comment]
On Thu 18 Jun, 2020, 5:50 AM Eric Schweitz, <notifications@github.com>
wrote:

> *@schweitzpgi* commented on this pull request.
> ------------------------------
>
> In flang/unittests/Optimizer/KindMappingTest.cpp
> <https://github.com/flang-compiler/f18-llvm-project/pull/184#discussion_r441901642>
> :
>
> > +
> +TEST_F(DefaultStringTests, getIntegerBitsizeTest) {
> +  EXPECT_EQ(defaultString->getIntegerBitsize(10), (unsigned)80);
> +  EXPECT_EQ(defaultString->getIntegerBitsize(0), (unsigned)0);
> +}
> +
> +TEST_F(DefaultStringTests, getCharacterBitsizeTest) {
> +  EXPECT_EQ(defaultString->getCharacterBitsize(10), (unsigned)80);
> +  EXPECT_EQ(defaultString->getCharacterBitsize(0), (unsigned)0);
> +}
> +
> +TEST_F(DefaultStringTests, getLogicalBitsizeTest) {
> +  EXPECT_EQ(defaultString->getLogicalBitsize(10), (unsigned)80);
> +  // Unsigned values are expected
> +  std::string actual = std::to_string(defaultString->getLogicalBitsize(-10));
> +  std::string expect("-80");
>
> Why is this test different than the others?
>

To check the results must be unsigned, if by mistake someone removes the
unsignedness that could be problem.
And for

getIntegerBitsize

getCharacterBitsize and

getLogicalBitsize

It's calling the same function under the hood.

So zero is anyways tested twice.

**Comment by Sameeranjoshi** [comment]
Rebased on top of latest fir-dev.

File: flang/unittests/Optimizer/KindMappingTest.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    delete clStringConflict;
+  }
+
+  KindMapping *commandLineString{};
+  KindMapping *clStringConflict{};
+  mlir::MLIRContext *context{};
+};
+
+TEST_F(DefaultStringTests, getIntegerBitsizeTest) {
+  EXPECT_EQ(defaultString->getIntegerBitsize(10), (unsigned)80);
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/KindMappingTest.cpp:53]: All of these (unsigned) casts can be replaced by making the literal unsigned.

Specifically, `(unsigned)80` can be written as `80u`.

File: flang/unittests/Optimizer/KindMappingTest.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+  KindMapping *defaultString{};
+  mlir::MLIRContext *context{};
+};
+
+struct commandLineStringTests : public testing::Test {
+public:
+  void SetUp() {
+    commandLineString = new KindMapping(
+        context, "i10:80,l3:24,a1:8,r54:Double,c20:X86_FP80,r11:PPC_FP128");
```
Comment (schweitzpgi) [inline flang/unittests/Optimizer/KindMappingTest.cpp:38]: Can you make the test use all of the floating-point semantic names for completeness?

**Comment by schweitzpgi** [comment]
These changes can be upstreamed to llvm-project as-is. Thanks, Sameeran.

But please make sure any changes (if requested) in Phabricator also get merged back here to keep the copies in synch.

**Pull request merged**